packages/payment/src/services/payment-provider.ts

```diff
@@ -0,0 +1,238 @@
+    this.paymentProviderRepository_ = container.paymentProviderRepository
+  }
+
+  @InjectTransactionManager("paymentProviderRepository_")
+  async create(
+    data: CreatePaymentProviderDTO[],
+    @MedusaContext() sharedContext?: Context
+  ): Promise<PaymentProvider[]> {
+    return await this.paymentProviderRepository_.create(data, sharedContext)
+  }
+
+  @InjectManager("paymentProviderRepository_")
+  async list(
+    @MedusaContext() sharedContext?: Context
+  ): Promise<PaymentProvider[]> {
+    return await this.paymentProviderRepository_.find(undefined, sharedContext)
+  }
+
+  retrieveProvider(providerId: string): IPaymentProcessor {
+    try {
+      return this.container_[`pp_${providerId}`] as IPaymentProcessor
+    } catch (e) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Could not find a payment provider with id: ${providerId}`
+      )
+    }
+  }
+
+  async createSession(
+    provider_id: string,
+    sessionInput: PaymentProcessorContext
+  ): Promise<PaymentProcessorSessionResponse["session_data"]> {
+    const provider = this.retrieveProvider(provider_id)
+
+    if (
+      !isDefined(sessionInput.currency_code) ||
+      !isDefined(sessionInput.amount)
+    ) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_ARGUMENT,
+        "`currency_code` and `amount` are required to create payment session."
+      )
+    }
+
+    const paymentResponse = await provider.initiatePayment(sessionInput)
+
+    if (isPaymentProcessorError(paymentResponse)) {
+      this.throwFromPaymentProcessorError(paymentResponse)
+    }
+
+    return (paymentResponse as PaymentProcessorSessionResponse).session_data
+  }
+
+  async updateSession(
+    paymentSession: {
+      id: string
+      data: Record<string, unknown>
+      provider_id: string
+    },
+    sessionInput: PaymentProcessorContext
+  ): Promise<Record<string, unknown> | undefined> {
+    const provider = this.retrieveProvider(paymentSession.provider_id)
+
+    const paymentResponse = await provider.updatePayment(sessionInput)
+
+    if (isPaymentProcessorError(paymentResponse)) {
+      this.throwFromPaymentProcessorError(paymentResponse)
+    }
+
+    return (paymentResponse as PaymentProcessorSessionResponse)?.session_data
+  }
+
+  async deleteSession(paymentSession: {
+    provider_id: string
+    data: Record<string, unknown>
+  }): Promise<void> {
+    const provider = this.retrieveProvider(paymentSession.provider_id)
+
+    const error = await provider.deletePayment(paymentSession.data)
+    if (isPaymentProcessorError(error)) {
+      this.throwFromPaymentProcessorError(error)
+    }
+  }
+
+  async createPayment(
+    data: CreatePaymentInput
+  ): Promise<Record<string, unknown>> {
+    const { payment_session, provider_id } = data
+    const providerId = provider_id ?? payment_session.provider_id
+
+    const provider = this.retrieveProvider(providerId)
+
+    /**
+     * NOTE: JUST RETRIEVE
+     */
```

Comment:
Do we need to revisit this?

Reply:
I pushed a [commit](https://github.com/medusajs/medusa/pull/6308/commits/e950a3227b42081cdf11184dd762cbe8a66272ab) where I removed `createPayment` API. It's not needed since `authorizeSession` will create payment.